### PR TITLE
Handle try in Flatten pass

### PR DESCRIPTION
**Pull request by aheejin**
This adds handling of try in the Flatten pass. This is in a way similar
to 'if' handling.

### CODE REVIEW

**Comment by aheejin** [comment]
This turned out not working in some cases yet; I'm working on a fix.

**Comment by aheejin** [comment]
This is ready for review. This has to land after #2583. Because #2583's code change is not reflected here, this does not pass tests now. I'll merge master after #2583 lands.

**Comment by aheejin** [comment]
I'll hold this for the moment, because we discovered a problem in #2583, the prerequisite for this.

**Comment by aheejin** [comment]
I updated this old PR to use the new EH spec. Now we can use #4348 and this is not dependent on #2583 anymore.

File: src/passes/Flatten.cpp
Comment (kripken) [inline src/passes/Flatten.cpp:388]: It looks like `handleBlockNestedPops` will scan for Trys without checking the features first. How about adding an early exit there if exceptions are not enabled?

File: src/passes/Flatten.cpp
Comment (kripken) [inline src/passes/Flatten.cpp:206]: I think this if just skips adding a set if the body is unreachable, as a minor optimization? I don't think it's needed IIUC. And the catch bodies do not do this check which makes it a little surprising to see it here.

File: src/passes/Flatten.cpp
Comment (kripken) [inline src/passes/Flatten.cpp:225]: if `prelude` is not null, isn't it equal to `tryy`? If so then we finalized it two lines above?

File: test/lit/passes/flatten-eh.wast
Comment (kripken) [inline test/lit/passes/flatten-eh.wast:103]: Please add a test for removing a try return value.

File: src/passes/Flatten.cpp
Comment (aheejin) [reply src/passes/Flatten.cpp:388]: Done at the beginning of `handleBlockNestedPops`.

File: src/passes/Flatten.cpp
Comment (aheejin) [reply src/passes/Flatten.cpp:225]: Hmm, I think you're right. I kind of copy-pasted `if`'s code... Should I fix `if` too?

Comment (aheejin) [reply src/passes/Flatten.cpp:225]: By the way, are these any different?
```cpp
expr->finalize();
```
vs.
```cpp
ReFinalize().visit(expr);
```

File: src/passes/Flatten.cpp
Comment (aheejin) [reply src/passes/Flatten.cpp:206]: I copy-pasted `if`'s code here again: https://github.com/WebAssembly/binaryen/blob/3c1c8ae829d0026024de55aa65dc3d49bd8fd37a/src/passes/Flatten.cpp#L152-L159

I tried to remove these `if (iff->ifTrue->type.isConcrete())` and `if (iff->ifFalse && iff->ifFalse->type.isConcrete())` to see if these also can be removed, but if I do that these test fail:
```
Failed Tests (2):
  Binaryen lit tests :: passes/flatten_simplify-locals-nonesting_souperify-single-use_enable-threads.wast
  Binaryen lit tests :: passes/flatten_simplify-locals-nonesting_souperify_enable-threads.wast
```

I haven't checked why they fail (and I have to run for today), but I guess there can be cases these are necessary...? If that's the case I think I should also check the catch bodies too. I'll do that for the moment, and if they turn out not to be necessary, I will remove them.

File: test/lit/passes/flatten-eh.wast
Comment (aheejin) [reply test/lit/passes/flatten-eh.wast:103]: Done

File: src/passes/Flatten.cpp
Comment (kripken) [reply src/passes/Flatten.cpp:206]: I see, thanks. In that case let's check the catch bodies too, that seems most consistent with the rest of the code.

File: src/passes/Flatten.cpp
Comment (kripken) [reply src/passes/Flatten.cpp:227]: Ah, let's fix if too then, looks like it's also redundant there. Probably a copy-paste error there from back in history...

Yes, those two expressions are identical in practice, but the first uses static type info to call the right finalize function while the second is dynamic.

**Review by kripken — APPROVED** [review]
lgtm % comments

**Comment by kripken** [comment]
Also please fuzz before landing.

File: src/passes/Flatten.cpp
Comment (aheejin) [reply src/passes/Flatten.cpp]: Removed the second (redundant) finalization from both `try` and `if`.

Comment (aheejin) [reply src/passes/Flatten.cpp]: Hmm, turns out `if`'s finalization was necessary, because `if` has `condition` so `prelude` can be different from `iff`. Restored `if`'s finalization.

File: src/passes/Flatten.cpp
Comment (aheejin) [reply src/passes/Flatten.cpp:205]: I think it's OK to remove these body checks (both in `if` and `try`); the reason those tests failed was not because they crashed, but because the filecheck results were different. As you said, this does a small optimization that prevents this form:
```wasm
(local.set $x
  (unreachable)
)
```

But I think it's OK to leave them (both in `if` and `try`) anyway; while `Flatten` doesn't try to do any optimizations itself, it wouldn't hurt, and maybe the output will be a little easier to read...

Comment (kripken) [reply src/passes/Flatten.cpp:205]: sgtm, this simple pattern doesn't add much code complexity and it does make the output smaller and simpler.

**Review by kripken — APPROVED** [review]
Nice!

**Comment by aheejin** [comment]
As we talked last week, the EH fuzzing is not really working atm, and I made a new PR that fuzzes EH (based on the initial contents), but it keeps finding new bugs now unrelated to Flatten. I'm not sure if holding off landing this until the fuzzer runs long enough is worth it; this doesn't add a new regression at least, because this feature has not been working anyway. How about landing this first?

**Comment by kripken** [comment]
Landing this now sgtm if it won't cause fuzz errors on `main`, which I guess it won't as EH fuzzing is still off as you said?

**Comment by aheejin** [comment]
Yes

**Pull request merged**
